RFC: Allow interfaces to implement other interfaces

mike-marcacci opened the pull request:
This fixes #295 by describing the way in which an interface may implement another interface. The rules are essentially identical to the rules governing implementation of an interface by an object. I'd be more than happy to champion this change (per the CONTRIBUTING.md process) so please direct any questions my way.

This is a pretty small change, but adds substantially to the expressiveness of GraphQL's type system.

If someone in the WG can give me a nod, I'll go ahead and implement it in [graphql-js](https://github.com/graphql/graphql-js).

Thanks in advance!

theorygeek commented on spec/Section 3 -- Type System.md:
```diff
@@ -553,11 +554,15 @@ of rules must be adhered to by every Object type in a GraphQL schema.
+         3. An object field type is a valid sub-type if it is an Interface type
+            and the interface field type is either an Interface type or a Union
+            type and the object field type is a possible type of the interface
+            field type.
```
I think this language would need to be tweaked slightly:

> An object field type is a valid sub-type if it is an Interface type and the interface field type is either an Interface type or a Union type and _every possible type of the_ object field type is _also_ a possible type of the interface field type.

For example:
```graphql
interface Pet {
  mother: Pet
}

interface Canine {
  breed: String
}

type Wolf implements Canine {
  breed: String
}

type Labrador implements Canine, Pet {
  breed: String

  # Invalid: One of the possible types of Canine (Wolf) is not a possible type of Pet
  mother: Canine
}
```

theorygeek replied:
The weird thing is that this rule (which seems necessary) would cause a brittle base class problem: You might add a new type to the schema that implements some interfaces, and not others, and suddenly it causes _some other part of the schema_ to be invalid.

vergenzt replied:
Doesn't the rule apply to this scenario as written? `Canine` is not a possible type of `Pet`, so `mother: Canine` is invalid in `Labrador`.

theorygeek replied:
I worded it that way because I wasn't thinking of an Interface type as being a possible type for another Interface type. But based on the language for [fragment spreads](http://facebook.github.io/graphql/October2016/#GetPossibleTypes()):

> If _type_ is an interface type, return the set of types implementing _type_

Then I think I'd actually agree with the language as written. Under the definition I proposed, this would be valid, but it would not be valid with the language in the PR:

```graphql
interface Pet {
  mother: Pet
}

interface Canine {
  breed: String
}

# +Pet interface
type Wolf implements Canine, Pet {
  breed: String
  mother: Wolf
}

type Labrador implements Canine, Pet {
  breed: String

  # Valid under my definition: Every possible type of Canine is also a possible type of Pet
  # Invalid under original: Canine _itself_ is not a possible type of Pet
  mother: Canine
}
```

So... I retract my suggestion 😄  

I think the wording as-is, assuming that the definition of "possible types" is the defined by [`GetPossibleTypes`](http://facebook.github.io/graphql/October2016/#GetPossibleTypes()) function in section 5.4.2.3, is better than what I was proposing.

leebyron commented:
I know this has been sitting for a while, but I think it has what it takes to move from Proposal to Draft.

alloy commented:
(@leebyron Slightly confused about the label, shouldn’t that then be `Draft (RFC 2)`?)

slightlytyler commented:
@mike-marcacci are you still championing this? I haven't been an attendee to WG meetings before but I'd like to help in anyway to move this forward. I can't think of a single change to GraphQL that would improve its use for my team and myself more than this

mike-marcacci commented:
Hi @slightlytyler - thanks for the note! I am (or am at least am intending to) continue pushing for this. I missed the last WG meeting because the agenda filename was [misnamed with the wrong date](https://github.com/graphql/graphql-wg/pull/133) 😑 but am certainly planning on bringing this up at the next meeting. It seems I have some new merge conflicts to fix here, which I'll make sure to address before then.

The biggest roadblock for this RFC has actually been getting the right eyes on [the js implementation I made last January](https://github.com/graphql/graphql-js/pull/1218). It's now all merge conflicts, so my main goal for the WG meeting is getting a commitment from maintainers to seriously consider the PR when I update it.

To be honest, I believe your appeal here – and any other votes of support – will help propel this feature forward. In past meetings the group's hesitance has not been due any technical criticism, but rather a general reluctance to introduce complexity without a clear demand. I don't totally agree that this actually adds complexity, but it is a change, and so I understand giving some pushback. So if you have a real-world use case that is made much better by this proposal, describing it here would go a long way in helping me argue for incorporation of this feature into the spec!

slightlytyler commented:
I can give a simple example which highlights a reoccurring issue I have encountered in developing GraphQL APIs for various applications. I'll give as real-world of an example as possible without over complicating it. 

For reference, I currently work at Docker on our distribution products. We use the concept of `Node`, `Edge`, `Connection` like many in the GraphQL world. We also try to use required types as much as possible, opting instead to use interfaces or unions to describe variants of a base type.

```graphql
interface Node {
    id: ID!
}

interface Edge {
    cursor: String!
    node: Node!
}

interface Connection {
    edges: [Edge!]!
    page: PageInfo!
}

interface Image {
    digest: String!
    size: Int!
    # ...
}

type SingleArchImage implements Image, Node {
    id: ID!
    digest: String!
    size: Int!
    layers: [Layer!]!
}

type MultiArchImage implements Image, Node {
    id: ID!
    digest: String!
    size: Int!
    references: [PlatformReference!]!
}

type ImageConnection implements Connection {
    edges: [ImageEdge!]!
    page: PageInfo!
}

type ImageEdge implements Edge {
    cursor: String!
    node: Image! # this breaks cus Image isn't a Node
}
```

The above schema is close to what I would consider an optimal model, only lacking the interface composition to make it work.

```graphql
interface Image implements Node {
    id: ID!
    digest: String!
    size: Int!
    # ...
}

type SingleArchImage implements Image {
    id: ID!
    digest: String!
    size: Int!
    layers: [Layer!]!
}

type MultiArchImage implements Image {
    id: ID!
    digest: String!
    size: Int!
    references: [PlatformReference!]!
}

# ...

type ImageEdge implements Edge {
    cursor: String!
    node: Image! # it should work?!
}
```

Let me know if this makes sense or if you require an clarification. If I can help you in anyway moving this forward please let me know

simPod commented:
I just stumbled upon this as well. Is there anything I can do to help to move this forward?

mike-marcacci commented:
There is finally a new WG meeting [on the schedule](https://github.com/graphql/graphql-wg/blob/master/agendas/2019-05-02.md), so I have rebased the PR.

@leebyron per [your comment above](https://github.com/graphql/graphql-spec/pull/373#issuecomment-426142468) can we get this moved to draft status?

michaelstaib commented:
Hi @mike-marcacci,

we spoke about this item at the last wg meeting and I want to get this one further.
Do you still want to champion this?

michaelstaib commented:
@mike-marcacci 

Here is the discussion we had on the wg

* Ivan: One concern is validation rules with fragments where there are already existing issues. We should create a prototype to gather feedback for a couple of months to gather feedback, shouldn’t keep it on stage two for now
* Lee: Getting details right is always difficult. Good to move slow and deliberately.
* Matt: As an example with validation rules, one piece of the interface proposal would be that we should tighten up validation rules, disallowing arbitrary spreading of interfaces. This effort is a breaking change for existing builds and should be staged.

the protocol is not yet approved but can be read in this pull request:
https://github.com/graphql/graphql-wg/pull/193/files

mike-marcacci commented:
Hi @michaelstaib - I am definitely committed to this. I had become a bit exasperated by the lack of participation: despite 18 months trying to [get some eyes on the implementation PR](https://github.com/graphql/graphql-js/pull/1218), no one stepped up. I am **thrilled** that there's renewed interest, and will definitely start pushing this again.

Updates since the last meeting in which I participated:

- I had to do a reworking of this RFC to fix many merge conflicts, so it needs to be reviewed again.
- The [implementation PR](https://github.com/graphql/graphql-js/pull/1218) needs to be rewritten to address merge conflicts in the dozens of test files I modified. Then it needs to receive its first review 😉.
- Any remaining conceptual concerns RE this RFC need to be discussed at a WG meeting. (I do think this has been discussed pretty extensively, though.)
- Any practical concerns RE this RFC need to be addressed at a WG meeting:
  - Interfaces will now appear in an introspection query in places that currently only have concrete types; how might this effect tooling?
  - Is there a conflict with arbitrary spreads here? I don't immediately see that there is anything particularly relevant in this PR, but perhaps I'm missing something. Either way, I'm generally very open to adopting stricter validation rules. One thing we should generally keep in mind while doing this, though, is not breaking the schema evolution practice in which a concrete type is changed into an interface (ie. `User` evolves from a concrete type to an interface implemented by `Bot`, `Individual`, and `Organization` concrete types).

I will add this to the next meeting's agenda and make sure I'm available to join. If there's any interest in discussing this or the proposed validation rules in an informal ad-hoc call, let me know and I'll make it happen.

🎉🎉🎉

mike-marcacci commented:
Copying those who participated in [the last WG discussion](https://github.com/graphql/graphql-wg/pull/193/files#diff-541802e783511df0621595d136665fe8R181) according to the (AMAZING) WG agenda notes: @michaelstaib @leebyron @IvanGoncharov @mjmahone

michaelstaib commented:
@mike-marcacci we could talk this through with @mjmahone and then have see what needs to be done in that area.

mike-marcacci commented:
Excellent – I can make any day this week work if we'd like to hop on a Google Hangout or Zoom call.

mjmahone commented:
Ah sorry I am catching up on my github notifs. I'd be happy to schedule something after this week's WG meeting, also happy to just listen at the WG to what's already been hashed out. Really happy to see progress!

michaelstaib commented:
@mike-marcacci good to see finally progress on this one 👍 I will start implementing this one in our GraphQL server implementation. I still think that the merging rules should be sufficient to handle the discussed edge cases... we'll see.

mike-marcacci commented on spec/Section 4 -- Introspection.md:
```diff
@@ -291,6 +289,7 @@ Fields
   * Accepts the argument `includeDeprecated` which defaults to {false}. If
     {true}, deprecated fields are also returned.
+* `interfaces`: The set of interfaces that an interface implements.
 * `possibleTypes` returns the list of types that implement this interface.
```
I finally got a chance to start reimplementing this in graphql-js (I'm just starting over, since there are a large number of conflicts and I want to give the entire codebase a scan for relevant changes anyhow.)

This has reminded me of one point I [briefly mentioned earlier](https://github.com/graphql/graphql-spec/pull/373#issuecomment-500554181) but want to call out here.

I intentionally left the `possibleTypes` field definition unchanged in the spec, which means that it's now possible for interfaces to show up in the list (as opposed to only concrete types). While this isn't technically problematic, I suspect that tooling has been written with assumptions that will no longer be true.

For anyone tracking this, I just want to bring attention to this minor "change cost."

andimarek replied:
We should consider alternatives, because I would expect that this breaks clients/tooling.

mike-marcacci replied:
I guess our options are:

1. Add interfaces to `possibleTypes`, which should technically be fine, since it's still a `__Type`, and the spec has never required a specific value for its `.kind`. Filtering for concrete types is as simple as ensuring `.kind` is `__TypeKind.OBJECT`.
2. Continue to only include implementing `__Type` objects where `.kind` is `__TypeKind.OBJECT`
    1. Don't do anything else. Finding relationships between interfaces is still possible, but considerably more challenging, as the tree must be traversed from leaves back to the trunk using the `.interfaces` field.
    2. Add a new field that includes only implementing interfaces.
    3. Add a new field that includes implementing interfaces and concrete types

Option 1 definitely feels the cleanest here... but I'm curious what kind of tools might have made this kind of assumption, and what kind of consequences it may have.

IvanGoncharov replied:
@mike-marcacci If you dump entire introspection you need `possibleTypes` only for union and since interfaces you can use `interfaces` field inside object type. An AFAIK it whats everyone currently doing, and here are examples in [graphql-js](https://github.com/graphql/graphql-js/blob/adb5292aea44baa989f3672d6db05f2949fbc0aa/src/utilities/buildClientSchema.js#L247),  [graphql-java](https://github.com/graphql-java/graphql-java/blob/3cbc1adfcf919970cb02fa8eb9a0cbc7fe1c3fab/src/main/java/graphql/introspection/IntrospectionResultToSchema.java#L216) and pretty sure all other implementations.

So I don't think we will actually break any tooling if we add interfaces to `possibleTypes`.

IvanGoncharov replied:
Also, in the absolute majority of cases, introspection queries are used by dev tools and breaking them is not such a big deal. Another argument would be that it's opt-in feature meaning it's up to API owners to evaluate if they breaks something or not by starting to use this feature.

mike-marcacci replied:
I agree – this is 100% my stance on this as well. I just wanted to call out the caveat here.

andimarek commented:
In general I am in favour of this change.

One thing I would like to see considered in the readability aspect with hierarchies of Interfaces:

If I understand the current proposal correctly you can do this:

```graphql
type Object implements SomeInterface{
    value: String
    node: ID!
}
interface SomeInterface implements Node {
  value: String
  node: ID!
}
interface Node {
  node: ID!
}
```
This means if I just look at the definition of `Object` I can't see anymore all Interfaces which `Object` implements. Before that it was always clear, now you can actually change the type signature of an Object without changing it.

The alternative would be to very explicit and require all Interfaces to be listed which are implemented by an Object even if they are inherited. For this example it would mean:

```graphql
type Object implements SomeInterface & Node {
   value: String
    node: ID!
} 
```
What do others think about that?

mike-marcacci commented:
Thanks for the comment @andimarek!

While thinking through the proposed changes I tried to find a balance that preserved the clarity that comes from explicitness against its unfriendliness of authorship. I considered your suggested approach, but ultimately came to the conclusion that very little clarity is lost by allowing interfaces to be "chained," provided each implementing interface fully defines a compatible superset of its implemented interfaces. The requirement to re-specify these "ancestor" interfaces felt unnecessary, given that changing hierarchies have no effect on field types, and only _where_ in a query a type might be returned.

As I write out this defense, however, I'm beginning to favor your approach. In particular, I suspect that very deep chains will be rare, so the ergonomic costs of your suggestion are very small in practice. For that, you gain some simplicity in tooling, a little bit of clarity, and (perhaps most importantly) prevent there from being two ways to do the same thing.

I dislike that in the current version of the RFC, **both** of these are valid, and functionally identical:
```
type Object implements SomeInterface {
  value: String
  node: ID!
}

type Object implements SomeInterface & Node {
  value: String
  node: ID!
} 
```

andimarek commented:
@mike-marcacci I would prefer to be very explicit and require all Interfaces to be named. This keeps the current semantics completely intact (you see all interfaces on the object directly) and enhances the readability aspect and minimises surprises. 

It is also a condition that could be relaxed in the future and allow indirectly implemented Interfaces to be omitted without being a breaking change. The other way around (be relaxed and then strict) would not be really possible because it would break existing SDLs.

mjmahone commented:
I also want to come down on the side of explicitness: to my knowledge, we had this same choice when choosing whether types needed to re-define the field that an interface they implements already defines. Forcing those to be explicitly defined lead to some nice properties, such as that you can resolve ambiguity and have things like:
```
interface HasNode {
  node: Node
}
type Tree implements HasNode {
  node: Leaf
}
type Leaf implements Node { ... }
```
While this is still possible in a world where fields don't *need* to be redefined, it means you either lose some type-locality (needing to look up what the interface's field value is), or you sometimes have over-defining (defining fields you don't need to). Given we've already chosen the "maximally explicit" paradigm, I think it'd be good to preserve that philosophy here.

mike-marcacci commented:
I'm completely in agreement here, and will add this language. Thanks for all of the very well articulated points!

IvanGoncharov commented:
@mike-marcacci @mjmahone @andimarek I don't agree that we should require types to list all interfaces it's indirectly implemented since it makes it harder to extend 3rd-party schema or reuse 3rd-party types.
For example, if you have this extension:
```graphql
extend type Query {
  foo: Foo
}

type Foo implements Node {
  id: ID!
}
```
and schema/libary author decides to extract `HasGlobalID` interface it will unnecessary brakes such extensions.
I think extracting common interfaces it pretty common refactoring and it shouldn't break extension or technology build on them (e.g. Schema Federation).

Also it makes impossible to extend the interface to implement new interfaces without enumerating all classes it implements:
```graphql
extend type Query {
  getFromCache(id: ID!): HasGlobalID
}

extend interface Node implements HasGlobalID

# Now you need to explicitly list all the types
extend type User implements HasGlobalID
extend type Company implements HasGlobalID
#... 
```

Also, I think it will be a big argument against creating chains of interfaces since after a certain number of interfaces since it will look too complicated. It will hurt future efforts to develop standardized interfaces (e.g. `HasGlobalID` or similar ones for connections, pagination, sorting, etc.) in the scope of single Schema or even community-wise (similar to Relay spec).

> This means if I just look at the definition of Object I can't see anymore all Interfaces which Object implements. Before that it was always clear, now you can actually change the type signature of an Object without changing it.

Every programming language that has interfaces has this problem (including Java) and is easily solved by compiler producing a user-friendly error or in case of GraphQL library `buidlSchema`/`extendSchema` functions can do the same.

Moreover, since GraphQL requires you to define all field you implement that means if you do such change signature of you may need to propagate this change to interfaces that you directly implement and repeat this process recursively.

> I also want to come down on the side of explicitness: to my knowledge, we had this same choice when choosing whether types needed to re-define the field that an interface they implements already defines.

It makes sense for fields since in this case, you can change something (field signature or attach some directives) but there is nothing you can change or override when you list interfaces you implement.

----

In general, for me, it's a question of encapsulation: when you implementing an interface you should care only about its list of fields and their signature.

IvanGoncharov commented:
I also forget to mention that by requiring to list all interfaces we actually lose information about author intention. 

And you can clearly see it in the documentation aspect. For example, if your organization has a standard interface for `Connection` in most cases it would be enough for the reader to just read it's description and list of fields but if we include all transitive interfaces read doesn't have any idea if also suppose to look into `RelayForwardConnection` and `RelayBackwardConnection`.

Same goes for tools like `graphql-voyager` where we will be forced to draw edges to all interfaces (since we don't know author intention) and this will significantly complicate graph that we need to draw.

So in general, by listing all interfaces we lose the ability to distinguish explicitly implemented interfaces and transitive interfaces.

mike-marcacci commented:
@IvanGoncharov these were the kind of observations that swayed me to initially make declaration of transitive interfaces optional. Thinking through your points here, we need to consider the impact on both schema authorship and on the introspection query.

It's also important to distinguish between effects on the ergonomics of authoring a schema from the ability to convey meaning through an introspection query.

If we care about the ability to express meaning through declaring (or not declaring) the implementation of an interface, we need to change the introspection query. The introspection query needs to provide a way to retrieve all implementations and all implemented interfaces in a single dimension (ie. not requiring recursive lookups), which would require the addition of new fields. Pehaps something like this would work:

- the `possibleTypes` field can list implementing types but not interfaces (both declared and transitive)
- a new `implementations` field can list all declared implementing types and interfaces of an interface
- a new `transitiveImplementations` field can list all transitive implementing types and interfaces of an interface
- the `interfaces` field can list declared implementations of a type or interface
- the `transitiveInterfaces` field can list transitively implemented interfaces

---

To avoid all this, in the current version of this RFC:

- the `possibleTypes` field lists all implementing types and interfaces (transitive and direct)
- the `interfaces` field lists all implemented interfaces (transitive and direct)

(While this is technically backwards-compatible and I believe this is a reasonable change to make, I've [started a discussion](https://github.com/graphql/graphql-spec/pull/373#pullrequestreview-262167133) about the possible consequences on tools that have made assumptions here.)

This eliminates the ability to convey any kind of special meaning through the selective declaration of transitive interfaces, since the result is normalized and flattened. This also makes the question of requiring declaration of transitive interfaces much less consequential, as we can require explicitness now and potentially relax the rules later.

This is my preferred approach.

IvanGoncharov commented:
@mike-marcacci If we require to explicitly list all implemented interfaces what are actual benefits of this feature against what you can already do with current spec:
```graphql
type Query {
  pet: Pet
  bestFriend: Canine
}

interface Pet {
  name: String
}

interface Canine {
  name: String
  breed: String
}

type Dog implements Pet & Canine {
  name: String
  breed: String
}
```

From both query validation/execution there is no difference if `Canine` implements `Pet` or not. So what are real benefits from adding it to the spec?

If you just want to enforce that all types that implement `Canine` should also implement `Pet` you can easily create custom `@implements` directive for that.

Personally, for me, this proposal was always about clearly expressing the author intends and help to expose it through introspection.
Without this motivation, I don't think this feature is worth added complexity since it doesn't bring any benefits to clients or client-side tooling comparing to what's already possible with current GraphQL specification.

mike-marcacci commented:
@IvanGoncharov while that certainly is a potential added benefit of this RFC (should we take that approach), for me this is mostly about allowing the type system to represent a case that was previously un-representable.

Currently there is no way to do the following, because without `interface Pet implements Node`, `PetEdge` is unable to implement `Edge`:


```graphql
# generic "utility" types
interface Node {
  id: ID!
}

interface Edge {
  cursor: String
  node: Node
}

# pet-specific types
interface Pet implements Node {
  id: ID!
  name: String
}

type Cat implements Pet & Node {
  id: ID!
  name: String
}

type Dog implements Pet & Node {
  id: ID!
  name: String
}

type PetEdge implements Edge {
  cursor: String
  node: Pet
}
```

Making this expressible is my primary motivation with this PR.

Adding the ability to indicate that `Cat` only implements `Node` _because it implements_ `Pet` (by omitting the explicit `implements Node` is something else that could be enabled here. However, as this RFC doesn't propose a way to communicate this via the introspection query, this is purely about ergonomics of authorship, and I don't have particularly strong feelings about that in either direction.

I'd like to suggest that maybe there isn't much to be communicated by omission of an explicit `implements`. I'm not completely sure about this, but as I try to think through use cases, it's easy to reassemble the hierarchy from the interface definitions themselves. From the example above with the type `Dog`, we _know_ that the implementation of `Pet` is more "important" than its implementation of `Node` because the latter is required by the former.

IvanGoncharov commented:
@mike-marcacci Thanks for the clarification 👍
I forgot about your initial use case so probably need to find some time to re-read the entire thread 😞 

Also, I forget to answer this one:
> If we care about the ability to express meaning through declaring (or not declaring) the implementation of an interface, then we need to change the introspection query so it's able to cary this information. Because an introspection query needs to provide a way to retrieve all implementations of an interface and all interfaces implemented by a type/interface in a single dimension (ie. without requiring recursive lookups), this would require the addition of new fields. 

It's not an issue if you dump entire introspection since you have values of `interfaces` for all types including all transitive interfaces. If we speak about partial introspection query 
I also don't see any problem here if:
1. the `possibleTypes` field lists all implementing types and interfaces (transitive and direct). This behavior logical and fits `possibleTypes` name since all transitive and direct types are possible values of the interface.
2. the `interfaces` field lists all implemented interfaces (only direct). But it can have `includeTransitive` boolean flag which is `false` by default. We already have very similar `includeDeprecated` flag on `fields` and `enumValues` fields.

Also to make introspection discussion more concrete, can you please describe your use case for "partial introspection" that needs to know this info?

leebyron commented:
Some thoughts on this (excellent, detailed) discussion:

* I think `possibleTypes` introspection field should only include `Object` types. The current spec language is 

  > possibleTypes returns the list of types that implement this interface. They must be object types.

  Same language for union types. This means the expected use of this field is to understand all possible types you might find from a field at runtime for a given abstract type.

  Because of this expected use, it should also definitely include object types found transitively.

* The `interfaces` introspection field should only include the interfaces directly implemented (regardless of the decision of whether object types should be required to directly implement an interface's implemented interfaces). 

  I'm going to call YAGNI on adding a `includeTransitive` argument unless there's a use case I've missed in the comment thread above.

* We should avoid adding additional fields (or arguments) to the introspection system if possible. It seems to me that we do not need anything new to represent this proposal, am I mistaken?

* Re: should interfaces be required to provide the fields of the interfaces they implement? 

  I agree with the consensus in this thread that yes, yes they should (otherwise we might have called this `extends` instead of `implements` - which I don't think we're moving towards). 

* Re: should object types be required to explicitly provide transitively implemented interfaces? 

  I do not have an opinion on this, and I can see the benefits on either side. For brevity let me refer to each as:

  Option A: objects being required to implement all transitively implemented interfaces
  Option B: objects only implement semantic interfaces, which can then implement other interfaces.

  In Option A, Seeing all interfaces per object in the SDL directly has readability benefits. You do not need to surf through interfaces to find the total set. This can be solved with a tool like GraphiQL in Object B, but it's always nice if the plain SDL is maximally readable without additional IDE tools.

  Option B also has readability and expressivity benefits. Specifying that a `Dog` is both a `Pet` and an `Animal` is maybe overly verbose when it is also the case that a `Pet` is an `Animal`.

  Option A seems like the minimal change to existing tools and implementations, since just ignoring the fact that interfaces can implement interfaces still yields a schema as expected. However existing tools would require explicit knowledge of this change for Option B if they do any processing on the relationship between objects and interfaces.

  If I channel my inner @dschafer, I would suggest Option A for now if only for the reason that it is more restrictive. That means a future change to the spec could enable Option B without creating a breaking change, however starting with Option B and later deciding Option A would have been better would not be possible to change without breaking some existing schema. However if we get consensus and strong rationale that Option B is the right thing to do (and tooling changes are minimal) then I'm happy to go with it.

mike-marcacci commented:
@IvanGoncharov you raise a very good point that my inclination toward flattening interface chains in introspection is based on ensuring a partial query is usable by itself; I have no real-world use cases for this, but just _assumed_ someone did. This seems to not be the case, so I totally agree with you and @leebyron that the `interfaces` field should just list explicitly defined interfaces. I also agree that `possibleTypes` should only include concrete types, including transitively implementing types (if we go with Option B).

I also agree that "Option A" is (only slightly) preferable, since it is easier to move from A to B than the other way around. I would like to point out, however, that by going with Option A we essentially eliminate the need for language about transitive types in introspection, since this requirement eliminates them. Following, while the change from A to B is _easier_ than B to A, it isn't exactly free of conflicts, as tooling will make assumptions about the completeness of these introspection fields, and the introduction of transitivity will break these.

leebyron commented:
> while the change from A to B is easier than B to A, it isn't exactly free of conflicts, as tooling will make assumptions about the completeness of these introspection fields, and the introduction of transitivity will break these.

Correct, but that's also true if we jumped straight to Option B from the current state of things. That's just an inherent complexity cost of Option B that we delay by starting with Option A and considering A to B as a future possible RFC. B to A on the other hand is actually a breaking change - so it's not just hard, it's probably not possible.

leebyron commented:
>  I would like to point out, however, that by going with Option A we essentially eliminate the need for language about transitive types in introspection, since this requirement eliminates them

Great point. This certainly feels like the minimal change - the only change to introspection is that interfaces may now support the interfaces field, otherwise nothing new.

leebyron commented on spec/Section 3 -- Type System.md:
```diff
@@ -816,11 +817,15 @@ of rules must be adhered to by every Object type in a GraphQL schema.
             type.
-         3. An object field type is a valid sub-type if it is a List type and
+         3. An object field type is a valid sub-type if it is an Interface type
+            and the interface field type is either an Interface type or a Union
```
A Union type can't be a subset of an Interface type - I think this needs to be more clear that the interface field type is an implemented Interface type. 

This is kind of a mouthful, I'll take a crack at it here but feel free to reword this and try to make this more clear:

An object field's type is a valid sub-type if it is an Interface type and the interface's field's type is an Interface implemented by the object field's type.

mike-marcacci replied:
I took another look at this, and ended up breaking out the "interface" language from item 2 (so that it now only deals with unions); I updated item 3 so that it deals with interface implementation by objects _and_ interfaces.

leebyron commented on spec/Section 3 -- Type System.md:
```diff
@@ -947,8 +952,8 @@ Object type extensions have the potential to be invalid if incorrectly defined.
-objects can then implement these interfaces which requires that the object type
-will define all fields defined by those interfaces.
+objects and interfaces can then implement these interfaces which requires that
+the object type will define all fields defined by those interfaces.
```
the ~object~ implementing type

leebyron commented on spec/Section 3 -- Type System.md:
````diff
@@ -1038,6 +1043,22 @@ interface. Querying for `age` is only valid when the result of `entity` is a
 }
 ```
 
+When defining an interface that implements another interface, the implementing
````
This might deserve it's own sub-section:

```
**Interfaces Implementing Interfaces**
```

mike-marcacci commented:
Perfect. I will make the language change and finish up implementation with this in mind!

leebyron commented on spec/Section 3 -- Type System.md:
```diff
@@ -1065,6 +1086,38 @@ Interface types have the potential to be invalid if incorrectly defined.
+            Object type and the implemented interface field type is either an
+            Interface type or a Union type and the implementing interface field
+            type is a possible type of the implemented interface field type.
+         3. An implementing interface field type is a valid sub-type if it is an
```
same feedback

leebyron commented on spec/Section 4 -- Introspection.md:
```diff
@@ -132,7 +132,7 @@ type __Type {
   fields(includeDeprecated: Boolean = false): [__Field!]
 
-  # should be non-null for OBJECT only, must be null for the others
+  # should be non-null for OBJECT and INTERFACE only, must be null for the others
```
For my own clarity - this means implementations will start returning `[]` instead of `null` for existing interfaces which do not implement others?

mike-marcacci replied:
This is correct, simply to mirror the behavior of OBJECT types

leebyron commented on spec/Section 5 -- Validation.md:
```diff
@@ -1131,8 +1131,8 @@ fragment catInDogFragmentInvalid on Dog {
 
-In scope of an object type, unions or interface spreads can be used
-if the object type implements the interface or is a member of the union.
+In scope of an object or interface type, unions or interface spreads can be
```
Rather than changing this, I'd suggest a new section. This section is about what spreads you may use in the context of an Object type.

The section after this is about what spreads you may use in the context of an Abstract type.

Perhaps there should be one more section about `Abstract Spreads in Abstract Scope`?

mike-marcacci replied:
This looks like a mistake of mine. I removed the "or interface", and applicable appropriate language already exists under "Abstract Spreads in Abstract Scope":

> Union or interfaces fragments can be used within each other. As long as there exists at least *one* object type that exists in the intersection of the possible types of the scope and the spread, the spread is considered valid.

mike-marcacci commented:
Thanks SO much for the thorough review @leebyron. I've updated the PR to address your points (hopefully with adequately parsable language). I also added a rule requiring explicit declaration of transitively implemented interfaces:

> If an object type declares that it implements one or more interfaces that implement other interfaces, it must also include each transitively implemented interface in its list of implemented interfaces.

mike-marcacci commented:
I've reimplemented this for graphql-js [in a new PR](https://github.com/graphql/graphql-js/pull/2084) that should be ready for review.

mike-marcacci commented:
I've added language and examples that forbid cyclic interface implementation. This ended up being trivial to write, thanks to the "flattening" of transitive implementations.

As far as I'm concerned, this PR is in its final state.

mjmahone commented:
> We make external schema extension very brittle so adding new interfaces to existing interfaces (without adding fields) is actually a breaking change now.

I think this can (and should!) be solved with tooling.

We should absolutely create something like an `extendExternalSchema()` in `graphql-js` that, given an external B that implements another C, adds a schema extension for each type/interface in your internal schema that implements B.

We've built something very similar to the above for interface extensions: when we have an interface that adds a client-only field, we have tooling that iterates over all object types implementing the interface, and adds that field too. The end result is a printed SDL that is  spec compliant, but that we can continue to update from the server (by passing the original "server SDL" back through the tooling pipeline). But only one person/team (me and my team) need to worry about this, whereas *every* team gets to see, explicitly, which fields exist on which  types.

IvanGoncharov commented on spec/Section 3 -- Type System.md:
```diff
@@ -1147,6 +1162,8 @@ Interface types have the potential to be invalid if incorrectly defined.
    implement other interfaces, it must also include each transitively
    implemented interface in its list of implemented interfaces.
+6. An interface cannot implement an interface that in turn implements the
+   original interface.
```
Since we decided that interfaces should implement all transitive interfaces, this means we can simplify this rule to `An interface cannot implement itself`.

mike-marcacci replied:
This is a great point; I've updated the language accordingly. One additional benefit of this wording is the symmetry with our implementation in graphql-js.

leebyron commented:
Working on taking this to Stage 3 today with a final editorial pass

https://github.com/graphql/graphql-wg/blob/master/notes/2019-11-07.md#intermediate-interfaces-rfc-blog-post-10m-mike

leebyron commented on spec/Section 3 -- Type System.md:
```diff
@@ -810,23 +811,25 @@ of rules must be adhered to by every Object type in a GraphQL schema.
+            is a possible type of the interface field type.
+         3. An object field type is a valid sub-type if it is an Object or
+            Interface type and the interface field type is an Interface type and
+            the object field type implements the interface field type.
```
I appreciate the clarity from breaking this rule out

leebyron commented on spec/Section 3 -- Type System.md:
```diff
@@ -835,6 +838,9 @@ of rules must be adhered to by every Object type in a GraphQL schema.
          must not be of a non-nullable type.
+5. If an object type declares that it implements one or more interfaces that
+   implement other interfaces, it must also include each transitively
+   implemented interface in its list of implemented interfaces.
```
This should be part of step 4 above. I think it can be stated more simply in a way that does not require transitive traversal (eg if each "implements" is valid, then they're all valid)

leebyron commented on spec/Section 3 -- Type System.md:
```diff
@@ -1072,6 +1134,42 @@ Interface types have the potential to be invalid if incorrectly defined.
+5. If an interface type declares that it implements one or more interfaces that
+   implement other interfaces, it must also include each transitively
+   implemented interface in its list of implemented interfaces.
+6. An interface cannot implement itself.
```
Repeated from rule 3 above

leebyron commented on spec/Section 3 -- Type System.md:
```diff
@@ -1072,6 +1134,42 @@ Interface types have the potential to be invalid if incorrectly defined.
+            (invariant) as the implemented interface field argument.
+      3. The implementing interface field may include additional arguments not
+         defined in the implemented interface field, but any additional argument
+         must not be required.
```
I may make an editorial change here to factor this out into an algorithm so that it can exactly mirror the same rules defined by object

leebyron commented on spec/Section 3 -- Type System.md:
````diff
@@ -1045,6 +1051,62 @@ interface. Querying for `age` is only valid when the result of `entity` is a
+}
+```
+
+Interface definitions must not contain cyclic references. This example is
````
This needs to be well defined in validation below

leebyron replied:
My mistake - this example explains what you can't do but it's well covered by the rule that an interface can't implement itself (also broken by this example) and that implementations must explicitly declare all interfaces they transitively inherit from (so there are no paths to follow)

leebyron commented on spec/Section 4 -- Introspection.md:
```diff
@@ -292,6 +292,7 @@ Fields
 * `fields`: The set of fields required by this interface.
   * Accepts the argument `includeDeprecated` which defaults to {false}. If
     {true}, deprecated fields are also returned.
+* `interfaces`: The set of interfaces that an interface implements.
```
```suggestion
* `interfaces`: The set of interfaces that this interface implements.
```

leebyron commented on spec/Section 5 -- Validation.md:
```diff
@@ -1133,7 +1133,7 @@ fragment catInDogFragmentInvalid on Dog {
 
 In scope of an object type, unions or interface spreads can be used
-if the object type implements the interface or is a member of the union.
+if the type implements the interface or is a member of the union.
```
technically the prior text was accurate (this is about object scope) but I see why you want to remove any chance for misreading

leebyron commented on spec/Section 5 -- Validation.md:
```diff
@@ -1173,8 +1173,8 @@ declaration, not its body.
 Union or interface spreads can be used within the context of an object type
-fragment, but only if the object type is one of the possible types of
-that interface or union.
+fragment or interface type fragment, but only if the type is one of the
```
I think this change is misplaced (maybe should go into "Abstract Spreads in Abstract Scope" below)

halkeye commented:
Does this being merged mean various languages are/can going to start supporting it?

mike-marcacci commented:
@halkeye yes, that is correct. This will be in `graphql-js@15`, which is currently out as a release candidate.

mike-marcacci commented:
For those following along here, [graphql-js v15](https://github.com/graphql/graphql-js/releases/tag/v15.0.0) was released today containing this feature. 🚀

During the many working group meetings we realized that this feature might be mistaken as enabling or encouraging a deeply hierarchal, classical-inheritance style of schema authorship. This is obviously not the goal and following such a pattern can make schemas difficult to evolve.

To address this, I wrote a [post on dev.to](https://dev.to/mikemarcacci/intermediate-interfaces-generic-utility-types-in-graphql-50e8) that explains the motivation behind this feature.

It would be really helpful if folks following this issue could re-familiarize themselves with these motivations, and then help guide those in their circles to the appropriate use of the feature.

Thanks team!

michaelstaib commented:
Great work @mike-marcacci :)

michaelstaib commented:
We have this now implemented and available in our version 11 preview for Hot Chocolate. This will really go a long way for many users.

pratiklodha commented:
Are there any official docs where interfaces implement other interfaces are shown and explained?

mannyluvstacos commented:
@pratiklodha , maybe this DEV.to link will serve?
https://dev.to/mikemarcacci/intermediate-interfaces-generic-utility-types-in-graphql-50e8

Mike Marcacci authored it :)

yaacovCR commented:
@mike-marcacci should unions also be allowed to implement interfaces for all the same reasons and in the exact same way?

https://github.com/graphql/graphql-spec/issues/518#issuecomment-799829609
